### PR TITLE
Add RegexGenerator to System.ComponentModel.TypeConverter.

### PR DESCRIPTION
**Pull request by Clockwork-Muse**
Added generator to source library, not to tests (no tests found)

### CODE REVIEW

**Comment by danmoseley** [comment]
Seems reasonable to me -- @stephentoub style OK for you?

Cool that this will be our first use in actual product.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerVerb.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerVerb.cs:13]: We shouldn't need to introduce a new type for this.  GetParameterReplacementRegex can just be a private partial static on DesignerVerb, which can also be marked partial; the partial has no affect on the compiled metadata.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerVerb.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerVerb.cs:40]: We shouldn't need to introduce a new helper for this.  Instead, `public DesignerVerb(string text, EventHandler handler) : base(handler, StandardCommands.VerbFirst)` can be changed to `public DesignerVerb(string text, EventHandler handler) : this(text, handler, StandardCommands.VerbFirst)`, and then all the logic remains in the other ctor.

File: src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj
Comment (stephentoub) [inline src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj:264]: Why is this needed?

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj:264]: See

https://github.com/dotnet/runtime/blob/cc2cb040cf3c6464269ca3b563ae4cfe705d45f7/eng/generators.targets#L9-L30

for how that works.

Comment (joperezr) [reply src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj:264]: Yeah, instead of having individual ProjectReferences on each project that uses it, we should instead update generators.targets to have all projects be able to consume the source generator.

**Comment by joperezr** [comment]
I understand it might be hard to do for every one of these PRs, but it would be cool to have a sense of the specific case perf gains  out of onboarding the source generator.

File: src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj
Comment (Clockwork-Muse) [reply src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj:264]: Ah, thanks.
I originally did it as a project reference because I ran across the dll one in a PR, before I saw the recommendations.  But yes, this would be much better.

Comment (Clockwork-Muse) [reply src/libraries/System.ComponentModel.TypeConverter/src/System.ComponentModel.TypeConverter.csproj]: I've changed to adding the targets to that file, and it works, but I'm not very experienced with some of the more arcane things in build files.  I know this is going to be reviewed anyways, but please be thorough.

File: eng/generators.targets
Comment (joperezr) [inline eng/generators.targets:80]: @eerhardt and I had some offline discussions about making source generators available to frameworks where they actually can't really be consumed outside of dotnet/runtime (which is the case for the Regex one) and I believe we landed on the idea that we shouldn't do this, and instead either: a) limit the source generator to only be used on platforms where it can be consumed externally, or b) expose it to other frameworks if we would see value for adding them.

If you agree, can we limit this change to only add the source generator in the netcoreapp case? (basically remove changes from line 75-80 here)

File: eng/generators.targets
Comment (joperezr) [inline eng/generators.targets:67]: @jkoritzinsky what is the reasoning behind having an explicit path and an implicit path for adding the DLLImport generator? Is that reason also applicable to the RegexGenerator? Basically my question is: for the regexgenerator should we just simplify and only enable it whenever EnableRegexGenerator property is set to true?

Comment (jkoritzinsky) [reply eng/generators.targets:67]: We needed the `ItemGroup`-based path to allow us to filter adding the generator ProjectReference based on referenced projects. With MSBuild static graph (which we use for our restore functionality), we can't add a `ProjectReference` in a target, so we're limited to evaluation time. Since properties are evaluated before items, we needed to do the generator reference add based on an item. The property was added for simplicity.

Once we aren't adding sources to builds (and all of the APIs the generator uses are approved) then we can significantly simplify this file.

**Comment by Clockwork-Muse** [comment]
I can simplify the targets file (and it has to be done anyways), but I took another look at this, and I'm not sure whether regex is even required here.

The pattern is `"\(\&.\)"` - note only the single `.` pattern - but the tests are only testing with `"(&.)"`, which makes it look like `.` should actually be escaped (possibly instead of `&`, because that character doesn't need escaping).  If so, that means the use of regex could be removed entirely, and a simple `string.Replace()` could be used instead.  I can't find any uses of `"(&.)"`, or even anything `"(&<anything>)"`, so I'm unsure if even the `Replace()` is required, but I'm not at all familiar with this part of the code base, and don't know what I should be looking for here.

I can find at least one identical usage in [reference source](https://source.dot.net/#System.Windows.Forms.Design/System/ComponentModel/Design/DesignerActionItem.cs,a089c4ec3989348e), but this file is **not** in this repository, so at least some use seems to be in outside packages.

Thoughts?

**Comment by danmoseley** [comment]
I was thinking maybe this is for strings containing an embedded keyboard shortcut, eg `C&opy Text` means that `o` is the keyboard shortcut for a menu item, and the text is `Copy Text`.  But then the regex would be `&` ... (hardly worth a regex)

@DustinCampbell perhaps you know about DesignerVerbs and what this might be for?

**Comment by danmoseley** [comment]
I'd be fine with merging as-is as the behavior is not changing..

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerVerb.cs
Comment (kasperk81) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesignerVerb.cs:32]:
```suggestion
        private static partial Regex GetParameterReplacementRegex();

        /// <summary>
```

File: eng/generators.targets
Comment (Clockwork-Muse) [reply eng/generators.targets:129]: I'm fine with removing "it" - what's "it" though, just the constant or that entire target?  That is, is that target there just to add that constant?

**Comment by DustinCampbell** [comment]
> I was thinking maybe this is for strings containing an embedded keyboard shortcut, eg `C&opy Text` means that `o` is the keyboard shortcut for a menu item, and the text is `Copy Text`. But then the regex would be `&` ... (hardly worth a regex)
> 
> @DustinCampbell perhaps you know about DesignerVerbs and what this might be for?

You've understood it correctly @danmoseley. The `&` marks the next character as a keyboard accelerator character when a `DesignerVerb` is surfaced in UI (such as in VS context menus).

**Comment by danmoseley** [comment]
@DustinCampbell to be clear then, the code could/should simply be `.Replace('&', '')`  with no regex involved (let alone the current broken one)

**Comment by Clockwork-Muse** [comment]
> let alone the current broken one

For clarity, broken because:
 - It also expects enclosing parenthesis 
 - It removes the accelerator character during the replace.

**Comment by danmoseley** [comment]
> > let alone the current broken one
> 
> For clarity, broken because:
> 
> * It also expects enclosing parenthesis
> * It removes the accelerator character during the replace.

I was thinking of both. Although it's so broken for that purpose one wonders how it could be. It's not the first time we look at some dusty corner of old code and it's not clear why something is some way.

If @DustinCampbell (from Winforms/designer world) thinks it's OK to just replace & with '' then I suggest we do that. We have months to discover if that breaks something.

**Comment by Clockwork-Muse** [comment]
> thinks it's OK to just replace & with '' then I suggest we do that.

I can't believe it took me this long to spot the problem with this: it fails if the ampersand is used "for itself" - If the text is (or should be) `"Build & Run Tests"`, for example.  Can we guarantee that we don't have any such uses, or won't?  Or is the verb guaranteed to be a single word?

File: eng/generators.targets
Comment (Clockwork-Muse) [reply eng/generators.targets:80]: I'm fine with this, but what effect is this going to have in other areas, like tests?  If a project has `net48` in `TargetFrameworks`, would that mean it couldn't be used there?  What about tests that have `PlatformDetection.IsNetFramework`?

Comment (eerhardt) [reply eng/generators.targets:80]: > If a project has net48 in TargetFrameworks, would that mean it couldn't be used there?

Correct. That behavior aligns with our customers' experience. The same goes for `netstandard2.0` and `net6.0`. Only `net7.0`+ targets should be able to consume the RegexGenerator.

Comment (Clockwork-Muse) [reply eng/generators.targets:80]: Right, so any project that is expected to reference .NetF _at all_ can't use the generator unless we get fancy with `#ifdef`s, then, which is something I'll have to consider for some of the other test projects.

Comment (stephentoub) [reply eng/generators.targets:80]: As an aside,  while technically at the moment the generated code would likely work downlevel, changes are coming that will end up using new API surface area in .NET 7, so the generator will only be usable with .NET 7+.

**Comment by maryamariyan** [comment]
Would be good to get sign off from memebers of @dotnet/area-infrastructure-libraries for this PR.

**Comment by eerhardt** [comment]
@joperezr @safern @carlossanlop - any thoughts here? I think this is just waiting for sign off from the libraries infrastructure team.

**Comment by joperezr** [comment]
I will take a look at this today. Apologies for missing the previous notification.

File: eng/generators.targets
Comment (ericstj) [reply eng/generators.targets:123]: I don't really see a scenario for this define or the target, I'd vote to remove them.

File: eng/generators.targets
Comment (ericstj) [reply eng/generators.targets:67]: @jkoritzinsky it sounds like you wanted to make it automatic always and you see the property as less desirable, where @joperezr was saying the property is simpler.  I do think a property is more WYSIWYG, and it seems acceptable for construction of the shared framework where we are more careful about layering/dependencies.

File: eng/generators.targets
Comment (joperezr) [inline eng/generators.targets:69]: Based on previous feedback by me and @ericstj, I believe we want to remove this implicit entrypoint and leave it to just the one above that depends on the property.

File: eng/generators.targets
Comment (joperezr) [inline eng/generators.targets:118]: As mentioned in a comment above, please do remove this target as we should use conditional compilation instead in case we needed this.

**Review by joperezr — COMMENTED** [review]
Left two small comments, but other than that this seems ready to go from my side. Thanks a lot @Clockwork-Muse for your patience and work on this!

**Comment by Clockwork-Muse** [comment]
Pushed changes as requested by @joperezr .

Currently, this PR performs no behavioral changes, simply switching to the regex generator.  However, there are still questions about what the correct behavior should be, and whether regex is even required in this situation.  Should I open an issue to have someone look at that?

**Comment by stephentoub** [comment]
> Should I open an issue to have someone look at that?

Sure.

File: eng/generators.targets
Comment (jkoritzinsky) [reply eng/generators.targets:67]: Yes, we wanted to make DllImportGenerator enabled automatically and the property was a fallback. We wanted to enable our generator automatically since we're consuming it throughout the runtime, including CoreLib. @stephentoub expressed a desire for us to enable it for all projects (including tests) and not even having a property for opt-in/out, but until we get all of our APIs approved, that's a little difficult.

File: eng/generators.targets
Comment (joperezr) [inline eng/generators.targets:80]:
```suggestion
          DependsOnTargets="ConfigureDllImportGenerator"
```

Since you removed the target, build is failing given this wasn't removed too.

**Comment by Clockwork-Muse** [comment]
Error just seems to be a build timeout?

**Comment by danmoseley** [comment]
Indeed. @dotnet/dnceng can you make anything of this? I can't find any more info
```
2022-02-08T20:26:47.5338090Z   Sending Job to OSX.1200.Amd64.Open...
2022-02-08T20:27:02.6974010Z   Sent Helix Job; see work items at https://helix.dot.net/api/jobs/fc4588bb-fb88-41d9-8528-65d9db3d9c28/workitems?api-version=2019-06-17
2022-02-08T20:27:03.7159050Z   Waiting for completion of job fc4588bb-fb88-41d9-8528-65d9db3d9c28 on OSX.1200.Amd64.Open
2022-02-08T22:51:52.0130490Z ##[error]The operation was canceled.
2022-02-08T22:51:52.0201180Z ##[section]Finishing: Send to Helix
```

**Comment by lpatalas** [comment]
@danmoseley We'll take all look.

**Comment by adiaaida** [comment]
This looks like it happened at one of the times where the os.1200.amd64.open queue was super backed up and had to be purged.

**Comment by MattGal** [comment]
> This looks like it happened at one of the times where the os.1200.amd64.open queue was super backed up and had to be purged.

It's definitely this, turning off core dumps in runtime unblocked it because System.Runtime.InteropServices test was just spending hours uploading dumps on every machine.

**Comment by lpatalas** [comment]
That makes sense. I created an issue to track the investigation here https://github.com/dotnet/core-eng/issues/15567 but I guess we can close it in this situation.

**Comment by danmoseley** [comment]
Thank you @lpatalas  @adiaaida

**Comment by Clockwork-Muse** [comment]
On main, the output item type has changed for the dll generator.  I'm assuming I should change things to match?

**Review by eerhardt — APPROVED** [review]
The changes in ComponentModel LGTM.

@dotnet/area-infrastructure-libraries - I don't see an approval there. Is everyone OK with the infra changes in this PR?

**Review by joperezr — APPROVED** [review]
Infra related changes look good to me.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**
